source/log_formatter.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/io/log_formatter.h>
+
+#include <aws/common/date_time.h>
+#include <aws/common/string.h>
+#include <aws/common/thread.h>
+
+#include <inttypes.h>
+#include <stdarg.h>
+
+/*
+ * Default formatter implementation
+ */
+
+// (max) strlen of "[<LogLevel>]"
+#define LOG_LEVEL_PREFIX_PADDING 7
+
+// (max) strlen of "[<ThreadId>]"
+#define THREAD_ID_PREFIX_PADDING 22
+
+// strlen of (user-content separator) " - " + "\n" + spaces between prefix fields + brackets around timestamp + 1 padding
+#define MISC_PADDING 10
+
+#define MAX_LOG_LINE_PREFIX_SIZE (LOG_LEVEL_PREFIX_PADDING + THREAD_ID_PREFIX_PADDING + MISC_PADDING + AWS_DATE_TIME_STR_MAX_LEN)
+
+struct aws_default_log_formatter_impl {
+    enum aws_date_format date_format;
+};
+
+static int s_default_aws_log_formatter_format_fn(
+    struct aws_log_formatter *formatter,
+    struct aws_string** formatted_output,
+    enum aws_log_level level,
+    aws_log_subject_t subject,
+    const char *format,
+    va_list args) {
+
+    (void)subject;
+
+    struct aws_default_log_formatter_impl *impl = (struct aws_default_log_formatter_impl *) formatter->impl;
+
+    if (formatted_output == NULL) {
+        return AWS_OP_ERR;
+    }
+
+    int result = AWS_OP_ERR;
+
+    /*
+     * Calculate how much room we'll need to build the full log line.
+     * You cannot consume a va_list twice, so we have to copy it.
+     */
+    va_list tmp_args;
+    va_copy(tmp_args, args);
```

Comment:
This seems pretty runtime wasteful, why not just allocate a decent chunk (like 1K) and log into that, allowing vsnprinf() to truncate?

Reply:
That's a good candidate for performance optimization but for now I'd like to not do any clipping yet and see how things fall out when actually measured.